### PR TITLE
Custom Loaders Support

### PR DESCRIPTION
**Pull request by kozlov721**
Added support for using custom loaders.

## Example

**source.py**
```python
from luxonis_train.utils.loaders import BaseLoaderTorch

class CustomLoader(BaseLoaderTorch):
    import torch
    from luxonis_ml.data import LabelType

    from luxonis_train.utils.loaders import LuxonisLoaderTorchOutput

    @property
    def input_shape(self) -> torch.Size:
        return self.torch.Size((1, 3, 256, 256))

    def __getitem__(self, _: int) -> LuxonisLoaderTorchOutput:
        from luxonis_ml.data import LabelType

        img = self.torch.rand(3, 256, 256)
        label = self.torch.zeros(1, 256, 256)
        label[0, 100:200, 100:200] = 1
        labels = {
            "default": {
                LabelType.SEGMENTATION: label,
            }
        }
        return img, labels

    def __len__(self) -> int:
        return 32

    def get_classes(self) -> dict[LabelType, list[str]]:
        return {self.LabelType.SEGMENTATION: ["square"]}
```

```bash
luxonis_train --source source.py \
              --config configs/segmentation_model.yaml \ 
              loader.name CustomLoader loader.params '{}'
```

## Breaking Config Changes

`dataset` section renamed to `loader`. Custom loader can be referenced by its name in the `LOADERS` registry. The default loader is `LuxonisLoaderTorch`. Former keys of the `dataset` field now go under `params` field in the `loader`. 

```yaml
loader:
  name: LuxonisLoaderTorch  # this is the default

  train_view: train
  val_view: val
  test_view: test

  params:
    dataset_name: coco_test
    bucket_storage: gcs

```

### CODE REVIEW

File: tests/unittests/test_core/test_archiver.py [hunk 1]
```diff
@@ -23,7 +25,8 @@
 HEAD_NAMES = [head_name for head_name in ImplementedHeads.__members__]
 
 
-class TestArchiver:
+@pytest.mark.skip()
```
Comment (conorsim) [inline tests/unittests/test_core/test_archiver.py:28]: Why do we skip this test?

**Review by conorsim — APPROVED** [review]
Looks like a nice improvement, thanks!

File: tests/unittests/test_core/test_archiver.py (hunk 1 above)
Comment (kozlov721) [reply tests/unittests/test_core/test_archiver.py:28]: The NN Archive in the latest dev of luxonis-ml is not compatible with the tests for now (we need to figure out the yolov8 keypoint outputs in the archive)

File: luxonis_train/core/core.py [hunk 2]
```diff
@@ -175,21 +154,26 @@ def __init__(
...
-            batch_size=self.cfg.trainer.batch_size,
-            num_workers=self.cfg.trainer.num_workers,
-            collate_fn=collate_fn,
-            drop_last=self.cfg.trainer.skip_last_batch,
-            sampler=sampler,
-        )
+                sampler = torch_data.WeightedRandomSampler(weights, num_samples)
+
+        self.pytorch_loaders = {
+            view: torch_data.DataLoader(
```
Comment (tersekmatija) [inline luxonis_train/core/core.py:160]: What if self.loaders[view] does not exist?

**Review by tersekmatija — COMMENTED** [review]
1/ Does this affect any of our current online datasets?
2/ Have we tested the kpts training?

Otherwise looks good, thanks!

**Comment by kozlov721** [comment]
> 1/ Does this affect any of our current online datasets?

No, this shouldn't change anything regarding the underlying datasets

>2/ Have we tested the kpts training?

Yes

**Comment by CaptainTrojan** [comment]
@kozlov721 What is the correct way of providing `BucketType` and `BucketStorage` to the default `LuxonisLoaderTorch`? The `DatasetConfig` ensured that strings like `internal`, `local, `gcs` etc. were mapped to the corresponding enum type.

**Pull request merged**